README.md
# Persian Mobile Date and Time picker

This library provides a component that can set year, month, day, hour, minute and second by sliding up or down.

- `Note`:
  There is no need to use the `moment-jalaali` or any other Jalali libraries and All details and functions are further explained in the [Date helpers](#date-helpers) section and if you need a function that we don't have, let us know to implement it.

### Demo

All functionalities and demos have documented here: [Live Demo](https://persian-tools.github.io/persian-mobile-datepicker)

## Theme

<div style="padding:30px;display:flex; flex-direction:row; justify-content: space-between">
    <img src="./images/theme.jpeg"  style="margin-top: 20px" />

</div>

## Getting Started

### Install

Using [npm](https://www.npmjs.com/):

    $ npm install --save @persian-tools/persian-mobile-datepicker

Using [yarn](https://yarnpkg.com):

    $ yarn add --save @persian-tools/persian-mobile-datepicker

### Import what you need

The following guide assumes you have some sort of ES2015 build set up using babel and/or webpack/browserify/gulp/grunt/etc.

1. Import functions, helpers and Types

```javascript
import {
  Picker,
  format,
  newDate,
  DatePickerConfig,
} from '@persian-tools/persian-mobile-datepicker';
```

2. Create Picker Configuration model

```javascript
const config: DatePickerConfig = {
  year: {
    caption: {
      text: 'سال',
    },
  },
  month: {
    caption: {
      text: 'ماه',
    },
  },
  day: {
    caption: {
      text: 'روز',
    },
  },
};
```

3. Use **Picker** component

```JSX
const App = () => {
  const [showPicker, setShowPicker] = React.useState(true);
  const [selectedDateValue, setSelectedDateValue] = React.useState();
  const [selectedDateEvents, setSelectedDateEvents] = React.useState([]);

  function handleSubmit(selectedDate) {
    const date = format(selectedDate.date, "d MMMM yyyy");
    const events = selectedDate.events;

    setSelectedDateValue(date);
    setSelectedDateEvents(events);
  }

  return <Picker
    isOpen={showPicker}
    config={config}
    minDate={newDate({ year: 1399, month: 9, day: 11 })}
    maxDate={newDate({ year: 1400, month: 1, day: 13 })}
    onSubmit={handleSubmit}
    onChange={handleSubmit}
    onClose={() => setShowPicker(false)}
    highlightWeekends
  />
  }
```

[Live Example](https://codesandbox.io/s/persian-mobile-datepicker-initialization-5o87t)

### Props

| Property               | Rquired | Type                                                                  | Description                                                                                                                       |
| :--------------------- | :------ | :-------------------------------------------------------------------- | :-------------------------------------------------------------------------------------------------------------------------------- |
| isOpen                 | Yes     | boolean                                                               | Picker open status                                                                                                                |
| theme                  | No      | string                                                                | Picker Theme                                                                                                                      |
| config                 | Yes     | [DatePickerConfig](#datepickerconfig)                                 | configuration of datepicker                                                                                                       |
| classNamePrefix        | No      | string                                                                | className of the datepicker                                                                                                       |
| initialValue           | No      | Date                                                                  | initial date of datepicker                                                                                                        |
| value                  | No      | [WheelPickerSelectEvent](#wheelpickerselectevent)                     | value of datepicker                                                                                                               |
| title                  | No      | string                                                                | title of datepicker                                                                                                               |
| onChange               | No      | (selected: [WheelPickerSelectEvent](#wheelpickerselectevent)) => void | Gets called when value of the picker changes                                                                                      |
| onSubmit               | No      | (selected: [WheelPickerSelectEvent](#wheelpickerselectevent)) => void | Triggered when you click on Submit button                                                                                         |
| onCancel               | No      | () => void                                                            | Call when user clicked on Cancel Button                                                                                           |
| onClose                | No      | () => void                                                            | Call when Picker Sheet modal has closed or User clicked on Cancel Button or User manually closed the Sheet modal by drag and drop |
| minDate                | No      | Date                                                                  | Specifies the minimum selectable day by user                                                                                      |
| maxDate                | No      | Date                                                                  | Specifies the maximum selectable day by user                                                                                      |
| endYear                | No      | number                                                                | The Minimum selectable year(Picker will calculate the `StartYear` by this approach: `currentYear` + `startYear`)                  |
| startYear              | No      | number                                                                | The Maximum selectable year(Picker will calculate the `StartYear` by this approach: `currentYear` + `startYear`)                  |
| addDayName             | No      | boolean                                                               | Add the name of the day of the week                                                                                               |
| highlightWeekends      | No      | boolean                                                               | Determines whether to mark weekend days with red or not. (weekend day is Friday)                                                  |
| highlightHolidays      | No      | boolean                                                               | Determines whether to mark holidays in day column.                                                                                |
| height                 | No      | number                                                                | Height of Picker Sheet modal                                                                                                      |
| submitText             | No      | string                                                                | Submit button text                                                                                                                |
| cancelText             | No      | string                                                                | Cancel button text                                                                                                                |
| showCancelButton       | No      | boolean                                                               | Display Cancel button                                                                                                             |
| disableSheetDrag       | No      | boolean                                                               | Disable drag for the sheet content                                                                                                |
| disableSheetHeaderDrag | No      | boolean                                                               | Disable drag for the sheet header                                                                                                 |

## Custom date unit

set `config` to configure year, month, day.

```javascript
config = {
  year: {
    caption: {
      text: 'سال',
    },
  },
  month: {
    caption: {
      text: 'ماه',
    },
  },
  day: {
    caption: {
      text: 'روز',
    },
  },
};
```

## datePickerConfig

```js
[key in DateConfigTypes]: DateConfigValuesModel

```

where the DateConfigTypes is one of `"year", "month", "day", "hour", "minute", "second"`

## Picker config Object Dto

| Property          | Type          | Description                                                                                                |
| :---------------- | :------------ | :--------------------------------------------------------------------------------------------------------- |
| caption           | Object        | an object with the props text `string` and style `CSSProperties` regular react style object                |
| formatter         | Function      | a function of [ PickerExtraDateInfo ](#PickerExtraDateInfo) to format every columns item text              |
| classname         | Function      | a function of [ PickerExtraDateInfo ](#PickerExtraDateInfo) for specifying the classNames                  |
| shouldRender      | Function      | a function of [ PickerExtraDateInfo ](#PickerExtraDateInfo) to specify which functionalities should render |
| columnStyle       | CSSProperties | the inline style of columns of datepicker                                                                  |
| itemStyle         | CSSProperties | the inline style of each of cell items                                                                     |
| selectedItemStyle | CSSProperties | the inline style of selected cell item                                                                     |

## WheelPickerSelectEvent

| Property | Type            | Description                                           |
| :------- | :-------------- | :---------------------------------------------------- |
| object   | PickerDateModel | an object of selected date, including year, month,... |
| events   | Array           | the array of events                                   |
| date     | Date            | the value of date                                     |

## PickerExtraDateInfo

| Property    | Type    | Description                                                                                                            |
| :---------- | :------ | :--------------------------------------------------------------------------------------------------------------------- |
| weekDay     | number  | the number of day of week                                                                                              |
| weekDayText | string  | can have these values: `شنبه , یک‌شنبه , دو‌شنبه , سه‌شنبه , چهار‌شنبه , پنج‌شنبه , جمعه` or other names if you prefer |
| monthText   | string  | is the text of months for e.g. `فروردین`                                                                               |
| dayOfYear   | number  | number of day in a year                                                                                                |
| isLeapYear  | boolean | whether the year is a leap year or not                                                                                 |
| isHoliday   | boolean | whether the day is a holiday or not                                                                                    |
| year        | number  | current year                                                                                                           |
| month       | number  | current month                                                                                                          |
| day         | number  | current day                                                                                                            |
| hour        | number  | current hour                                                                                                           |
| minute      | number  | current minute                                                                                                         |
| second      | number  | current second                                                                                                         |

## Usage Examples

### Changing the months text

You can do it with the help of formatter props of month in config

```javascript
{
  year: {
    caption: {
      text: "سال",
    }
  },
  month: {
    caption: {
      text: "ماه",
    },
    formatter: ({ month, monthText }) => (month === 5 ? "امرداد" : monthText),
  },
  day: {
    caption: {
      text: "روز",
    }
  }


```

<div style="padding:30px">
  <img src="./images/monthFormatter.png" width="300" />
</div>

### Removing a day

If for whatever reason you want to remove some cells you can do so with the use of `shouldRender`

the following config code will remove the 6th day of 3rd month

```javascript
{
  year: {
    caption: {
      text: "سال",
    }
  },
  month: {
    caption: {
      text: "ماه",
    }
  },
  day: {
    caption: {
      text: "روز",
    },
    shouldRender: ({ month, day }) => !(month === 3 && day === 6),
  }

```

<div style="padding:30px">
  <img src="./images/shouldRenderer.png" width="300" />
</div>

### Styling columns

If you want to change the inline style of columns or cells of datepicker you can use `columnStyle`, `itemStyle` and `selectedItemStyle`

```javascript
{
  year: {
    caption: {
      text: "سال",
    },
  },
  month: {
    caption: {
      text: "ماه",
    },
    columnStyle: { background: "#aaa", color: "green" },
    itemStyle: { color: "green" },
    selectedItemStyle: { color: "blue" },
  },
  day: {
    caption: {
      text: "روز",
    },
  }

```

<div style="padding:30px">
  <img src="./images/columnStyle.png" width="300" />
</div>

### Date helpers

- `newDate`: Convert Jalaali Date to Gregorian and returns a Date instance

```javascript
import { newDate } from '@persian-tools/persian-mobile-datepicker';

newDate({
  year: 1400,
  month: 1,
  day: 1,
}); // Sun Mar 21 2021 00:00:00 GMT+0330 (Iran Standard Time)
```

- `convertDateInstanceToDateObject`: Convert entered date to an object

```javascript
import { convertDateInstanceToDateObject } from '@persian-tools/persian-mobile-datepicker';

convertDateInstanceToDateObject(new Date()); // { year: 1400, month: 5, day: 15, hour: 22, minute: 20,second: 10 }
```

- `daysInMonth`: Get the number of days in a month of a year

```javascript
import { daysInMonth } from '@persian-tools/persian-mobile-datepicker';

daysInMonth(1400, 1); // 31
daysInMonth(1399, 12); // 30 -> 1399 is a leap year
daysInMonth(1400, 12); // 29
```

- `getWeekDay`: Get the day of the week of the given date. Returns number starts from 0, 0 means the first day of Week and 6 means the last day of Week

```javascript
import { getWeekDay } from '@persian-tools/persian-mobile-datepicker';

getWeekDay(1400, 5, 15); // 6
```

- `getDayOfYear`: Get the day of the year of the given date.

```javascript
import { getDayOfYear } from '@persian-tools/persian-mobile-datepicker';

getDayOfYear(1400, 5, 15); // 139
```

- `isWeekend`: Return true if the Date is at the Weekend

```javascript
import { isWeekend } from '@persian-tools/persian-mobile-datepicker';

isWeekend(1400, 5, 22); // true -> because it is friday
isWeekend(1400, 5, 23); // false -> it is Saturday
```

- `getWeekDayText`: Get Name the days of the week

```javascript
import { getWeekDayText } from '@persian-tools/persian-mobile-datepicker';

getWeekDayText(1400, 5, 22); // جمعه
getWeekDayText(1400, 5, 23); // شنبه
```

- `isValidJalaaliDate`: Is the given Jalaali date valid?

```javascript
import { isValidJalaaliDate } from '@persian-tools/persian-mobile-datepicker';

isValidJalaaliDate(1399, 12, 30); // true -> because 1399 is leap
isValidJalaaliDate(1400, 12, 30); // false
```

- `isBefore`: Is the first date before the second one?

```javascript
import { isBefore, newDate } from '@persian-tools/persian-mobile-datepicker';

isBefore(newDate(1399, 12, 30), newDate(1400, 2, 1)); // true
isBefore(newDate(1400, 2, 1), newDate(1399, 12, 30)); // false
```

- `isAfter`: Is the first date after the second one?

```javascript
import { isAfter, newDate } from '@persian-tools/persian-mobile-datepicker';

isAfter(newDate(1361, 10, 10), newDate(1372, 10, 10)); // false
isAfter(newDate(1372, 10, 10), newDate(1361, 10, 10)); // true
```

- `format`: Returns the formatted date string in the given format. The result may vary by locale.
[See Accepted patterns](https://date-fns.org/v2.23.0/docs/format)
```javascript
import { format } from '@persian-tools/persian-mobile-datepicker';

format(new Date('2020-10-10'), 'yyyy/MM/dd'); // 1399/07/19
```

- `isEqual`: Are the given dates equal?

```javascript
import { isEqual, newDate } from '@persian-tools/persian-mobile-datepicker';

isEqual(newDate(1361, 10, 10), newDate(1372, 10, 10)); // false
isEqual(newDate(1361, 10, 10), newDate(1361, 10, 10)); // true
```

- `getCurrentYear`: Get the year of the current date.

```javascript
import { getCurrentYear } from '@persian-tools/persian-mobile-datepicker';

getCurrentYear(); // 1400
```

- `currentDateObject`: Converts date instance to an object.

```javascript
import { currentDateObject } from '@persian-tools/persian-mobile-datepicker';

currentDateObject(); // {day: 16, hour: 5, minute: 51, month: 5, second: 1, year: 1400}
```

- `isLeapYear`: Check if the entered year is Leap

```javascript
import { isLeapYear } from '@persian-tools/persian-mobile-datepicker';

isLeapYear(1399); // true
isLeapYear(1400); // false
```

## Roadmap

- [ ] Add hour, minute and seconds
- [ ] Implement react native component


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.13](https://github.com/persian-tools/persian-mobile-datepicker/compare/v0.0.12...v0.0.13) (2021-08-14)


### Bug Fixes

* bundle installation package-lock size ([9b4237b](https://github.com/persian-tools/persian-mobile-datepicker/commit/9b4237b7a2095abbb14a9ff99e10d30746b1ab0b))


### Docs

* **README:** update WheelPickerSelectEvent description ([5b0e217](https://github.com/persian-tools/persian-mobile-datepicker/commit/5b0e21760de2dd41dc7bf0d9eb07f574965c68d4))

### [0.0.12](https://github.com/persian-tools/persian-mobile-datepicker/compare/v0.0.11...v0.0.12) (2021-08-09)


### Docs

* **README:** add accepted format patterns ([a7138d7](https://github.com/persian-tools/persian-mobile-datepicker/commit/a7138d70000723e4f0b065ff75e242e4f8d43624))
* **README:** fix isAfter and isBefore docs ([e57d986](https://github.com/persian-tools/persian-mobile-datepicker/commit/e57d986a8fd2d7012f676b56c9458a64c1541b8c))

### [0.0.11](https://github.com/persian-tools/persian-mobile-datepicker/compare/v0.0.10...v0.0.11) (2021-08-09)


### Docs

* **storybook:** fix storybook source code loader ([2ba5d9e](https://github.com/persian-tools/persian-mobile-datepicker/commit/2ba5d9e8d2bff29559e70eb92dca89c8b99fdffb))

### [0.0.10](https://github.com/persian-tools/persian-mobile-datepicker/compare/v0.0.9...v0.0.10) (2021-08-09)


### Features

* categorize accessibility stories ([a41ca9c](https://github.com/persian-tools/persian-mobile-datepicker/commit/a41ca9c0f0c53b37cfbc2015b4caf40c75838820))


### Docs

* **storybook:** groups order and add introduction page ([3500947](https://github.com/persian-tools/persian-mobile-datepicker/commit/350094728879a776a6eedaa777991a4c131a0189))
* **storybook:** update all filter stories ([4c70764](https://github.com/persian-tools/persian-mobile-datepicker/commit/4c707645c096ca5b3518f0b9b226caa3487e6a46))
* **storybook:** update all styling stories ([8c44616](https://github.com/persian-tools/persian-mobile-datepicker/commit/8c44616136d28ce33173780c54d80d7221cdde4a))

### [0.0.9](https://github.com/persian-tools/persian-mobile-datepicker/compare/v0.0.11...v0.0.9) (2021-08-09)

### Bug Fixes

* highlightHolidays issue ([ab56cd8](https://github.com/persian-tools/persian-mobile-datepicker/commit/ab56cd86f1fabab1914460fc013441c5e664895e))

### Minor Improvements

* add sideEffect to false in package.json file ([1539664](https://github.com/persian-tools/persian-mobile-datepicker/commit/15396645d55e0f177d0f10e9128d2d27475cfa7b))
* add versionrc file ([da80d33](https://github.com/persian-tools/persian-mobile-datepicker/commit/da80d3330a37c6d43bf30005164ffe7555488e8d))

### Docs

* **README:** change them image ([702cb40](https://github.com/persian-tools/persian-mobile-datepicker/commit/702cb40e0f706cb5eacf10c6877a702216c56cd2))
* **README:** fix typo and add more info ([17c740b](https://github.com/persian-tools/persian-mobile-datepicker/commit/17c740b4cccc67351698eff689f98c4f46e2490a))
* **storybook:** remove console addons ([5b87482](https://github.com/persian-tools/persian-mobile-datepicker/commit/5b874826ee17b362c11e392472d5a10bf2d0d999))
* **storybook:** update all Columns Text formatter stories and fix the Stories source code ([5d90197](https://github.com/persian-tools/persian-mobile-datepicker/commit/5d90197ee7007b82bdb7705cb31593cff30ebe4e))
* **storybook:** update all Date range stories and fix the Stories source code ([1a41204](https://github.com/persian-tools/persian-mobile-datepicker/commit/1a412046218a54d1c6d8f9f68d3505892931108e))
* **usePicker:** add more docs for defaultSelectedDateObject complex logics ([7a7078a](https://github.com/persian-tools/persian-mobile-datepicker/commit/7a7078ac1c87df19f80409331048f6f407b494cb))

### [0.0.8](https://github.com/persian-tools/persian-mobile-datepicker/compare/v0.0.7...v0.0.8) (2021-08-08)

### [0.0.7](https://github.com/persian-tools/persian-mobile-datepicker/compare/v0.0.6...v0.0.7) (2021-08-08)


### Bug Fixes

* **usePicker:** min and max date issue ([0bd3631](https://github.com/persian-tools/persian-mobile-datepicker/commit/0bd363112e1b8dce52682f35055bd53140e0fd07))

### [0.0.6](https://github.com/persian-tools/persian-mobile-datepicker/compare/v0.0.5...v0.0.6) (2021-08-07)


### Bug Fixes

* build issue ([21f815d](https://github.com/persian-tools/persian-mobile-datepicker/commit/21f815dab8d2d393bd20f892ca6fd0beb2925bb2))

### [0.0.5](https://github.com/persian-tools/persian-mobile-datepicker/compare/v0.0.4...v0.0.5) (2021-08-07)


### Features

* **README:** add demo url and some notes ([e267367](https://github.com/persian-tools/persian-mobile-datepicker/commit/e267367b5aa3f9911bae532d0e2d785f8e3fac26))


### Bug Fixes

* **npm:** add source map files and stories to npmignore file ([25e759b](https://github.com/persian-tools/persian-mobile-datepicker/commit/25e759b3c2f1c4a0402008b43ef42d7369ec3432))

### [0.0.4](https://github.com/persian-tools/persian-mobile-datepicker/compare/v0.0.4-beta.0...v0.0.4) (2021-08-07)

### [0.0.4-beta.0](https://github.com/persian-tools/persian-mobile-datepicker/compare/v0.0.3...v0.0.4-beta.0) (2021-08-06)


### Features

* add documentations ([3c48e16](https://github.com/persian-tools/persian-mobile-datepicker/commit/3c48e164728ce5949c9ed20252b1a602309ba878))
* add roadmap ([b18bea6](https://github.com/persian-tools/persian-mobile-datepicker/commit/b18bea6bb4a2d51f28ab51f5e53e00d4e41d0ec7))
* added onCancel click event and minor improvements ([04a7de4](https://github.com/persian-tools/persian-mobile-datepicker/commit/04a7de4a60feb818c36aead17ae63737acf377a9))
* improve document ([6763585](https://github.com/persian-tools/persian-mobile-datepicker/commit/676358561c7d7404eec466e28f45494d085eb8f9))

src/helpers/date.ts
// Date Utilities
import {
  format as formatFns,
  getDay as dateGetDay,
  newDate as newDateFns,
  isValid as dateIsValid,
  isLeapYear as dateIsLeapYear,
  getDaysInMonth as dateGetDaysInMonth,
  getDate,
  getYear,
  getMonth,
  isEqual as isEqualFns,
  isBefore as isBeforeFns,
  isAfter as isAfterFns,
  getDayOfYear as getDayOfYearFns,
} from 'date-fns-jalali';
// @ts-ignore
import { isValidJalaliDate as isValidJalaliDateFns } from 'date-fns-jalali/_jalali/index';
// Helpers
import {
  generateAnArrayOfNumbers,
  generateArrayInRangeOfNumbers,
  toPositive,
} from './';
// Types
import type {
  PickerItemModel,
  WeekDayText,
  PickerDateModel,
  RequiredPickerDateModel,
} from '../components/WheelPicker/index.types';

export const weekDays: Record<number, WeekDayText> = {
  0: 'شنبه',
  1: 'یک‌شنبه',
  2: 'دو‌شنبه',
  3: 'سه‌شنبه',
  4: 'چهار‌شنبه',
  5: 'پنج‌شنبه',
  6: 'جمعه',
};
export const jalaliMonths = {
  1: 'فروردین',
  2: 'اردیبهشت',
  3: 'خرداد',
  4: 'تیر',
  5: 'مرداد',
  6: 'شهریور',
  7: 'مهر',
  8: 'آبان',
  9: 'آذر',
  10: 'دی',
  11: 'بهمن',
  12: 'اسفند',
};

/**
 * Converts Jalaali Date to Gregorian
 *
 * @public
 */
export function newDate({ year, month, day = 1 }: PickerDateModel): Date {
  // date-fns month starts from 0, it means farvardin is 0 and esfand is 11
  month = month! - 1;

  return newDateFns(year!, month, day, 0, 0, 0);
}

/**
 * Converts entered date to an object
 *
 * @public
 */
export const convertDateInstanceToDateObject = (
  date: Date,
): RequiredPickerDateModel => {
  return {
    year: getYear(date),
    month: getMonth(date) + 1,
    day: getDate(date),
    hour: 0,
    minute: 0,
    second: 0,
  };
};

/**
 * Get the number of days in a month of a year
 *
 * @public
 */
export function daysInMonth(year: number, month: number): number {
  return dateGetDaysInMonth(newDate({ year, month, day: 1 }));
}

/**
 * Get the day of the week of the given date.
 *
 * @description Returns number, starts from 0, 0 means the first day of Week and 6 means the last day of Week
 * @public
 */
export function getWeekDay(year: number, month: number, day: number): number {
  return (dateGetDay(newDate({ year, month, day })) + 1) % 7;
}

/**
 * Get the day of the year of the given date.
 *
 * @public
 */
export function getDayOfYear(year: number, month: number, day: number): number {
  return getDayOfYearFns(newDate({ year, month, day }));
}

/**
 * Returns true if the Date is at the Weekend
 *
 * @public
 */
export function isWeekend(year: number, month: number, day: number): boolean {
  return getWeekDay(year, month, day) === 6;
}

/**
 * Get Name the days of the week
 *
 * @public
 */
export function getWeekDayText(
  year: number,
  month: number,
  day: number,
): WeekDayText {
  const result = getWeekDay(year, month, day);

  return weekDays[result];
}

/**
 * Is the given date valid?
 *
 * Returns false if argument is Invalid Date and true otherwise.
 * Argument is converted to Date using toDate.
 * See toDate Invalid Date is a Date, whose time value is NaN.
 *
 * @public
 */
export function isValid(date: Date): boolean {
  return dateIsValid(date);
}

/**
 * Is the given Jalali date valid?
 *
 * @public
 */
export function isValidJalaaliDate(
  jy: number,
  jm: number,
  jd: number,
): boolean {
  return isValidJalaliDateFns(jy, jm, jd);
}

/**
 * Is the first date before the second one?
 *
 * @public
 */
export function isBefore(dateLeft: Date, dateRight: Date): boolean {
  return isBeforeFns(dateLeft, dateRight);
}

/**
 * Is the first date after the second one?
 *
 * @public
 */
export function isAfter(dateLeft: Date, dateRight: Date): boolean {
  return isAfterFns(dateLeft, dateRight);
}

/**
 * Return the formatted date string in the given format. The result may vary by locale.
 *
 * > ⚠️ Please note that the `format` tokens differ from Moment.js and other libraries.
 * > See: https://git.io/fxCyr
 *
 * @public
 */
export function format(date: Date | number, formatBy: string): string {
  return formatFns(date, formatBy);
}

/**
 * Are the given dates equal?
 *
 * @public
 */
export function isEqual(
  dateLeft: Date | number,
  dateRight: Date | number,
): boolean {
  return isEqualFns(dateLeft, dateRight);
}

/**
 * Returns the current Year
 *
 * @public
 */
export function getCurrentYear(): number {
  return currentDateObject().year;
}

export type CurrentDateObject = Required<Record<keyof PickerDateModel, number>>;
/**
 * Converts date instance string to an object.
 * The result may vary by locale.
 *
 * @public
 */
export function currentDateObject(): CurrentDateObject {
  const date = new Date();
  return {
    year: Number(format(date, 'yyyy')),
    month: Number(format(date, 'M')),
    day: Number(format(date, 'd')),
    hour: Number(format(date, 'h')),
    minute: Number(format(date, 'm')),
    second: Number(format(date, 'S')),
  };
}

/**
 * Check if the entered year is Leap
 *
 * @public
 */
export function isLeapYear(year: number): boolean {
  return dateIsLeapYear(newDate({ year, month: 1, day: 1 }));
}

/**
 * Calculate the minimum year items which picker should render in Year Column
 *
 * @public
 * @example
 * assume that currentYear is 1400
 * startYear(1380) // 20(currentYear - 1380)
 */
export function startYear(year: number) {
  const currentYear = getCurrentYear();
  if (currentYear / year > 2) {
    throw new Error(
      `[PersianMobileDatePicker] Invalid Year, Usage: startYearFrom(1380), means Year picker's column should starts from 1380 up to now`,
    );
  }

  return toPositive(currentYear - year);
}

/**
 * Calculate the maximum year items which picker should render in Year Column
 *
 * @public
 * @example
 * assume that currentYear is 1400
 * endYear(1410) // 10(year - currentYear)
 */
export function endYear(year: number) {
  const currentYear = getCurrentYear();
  if (currentYear / year > 2) {
    throw new Error(
      `[PersianMobileDatePicker] Invalid Year, Usage: endYearTo(1410), means Year picker's column should end in 1410`,
    );
  }
  return toPositive(currentYear - year);
}

/**
 * Combines and Generates all picker columns value
 *
 * @private
 */
export const pickerData: Record<string, (inp?: any) => Array<PickerItemModel>> =
  {
    getYears: ({ min, max } = {}) =>
      generateArrayInRangeOfNumbers(min, max).map((year) => ({
        value: year,
        type: 'year',
      })),
    getMonths: (monthsMap = jalaliMonths) =>
      Object.keys(monthsMap).map((value) => ({
        type: 'month',
        value: Number(value),
      })),
    getDays: (days = 29) =>
      generateAnArrayOfNumbers(days).map((day) => ({
        value: day,
        type: 'day',
      })),
    getHours: () =>
      generateAnArrayOfNumbers(24).map((hour) => ({
        value: hour,
        type: 'hour',
      })),
    getSeconds: () =>
      generateAnArrayOfNumbers(59).map((hour) => ({
        value: hour,
        type: 'second',
      })),
    getMinutes: () =>
      generateAnArrayOfNumbers(59).map((hour) => ({
        value: hour,
        type: 'minute',
      })),
  };
